_posts/2021-08-31-modifier-type-clef-primaire-non-referencee-vers-biginit.md

```diff
@@ -0,0 +1,257 @@
+
+Nous avons prévu de migrer d'autres tables, référencées, en charge nominale. Nous partagerons les résultats dans un autre article.
+
+## Motivation
+
+En un an, depuis septembre 2020, la fréquentation de Pix a augmenté [de manière spectaculaire](https://pix.fr/statistiques). La charge a été bien absorbée par l'infrastructure, mais nous nous sommes rendu compte d'une autre conséquence: une table allait épuiser l'identifiant utilisé par sa clef primaire.
+
+Il est bien sûr possible de modifier le type de cette propriété avec l'instruction native `ALTER COLUMN TYPE`, mais cela a un coût proportionnel au remplissage de la table. Une exécution de cette instruction sur les données de production, sur un environnement moins puissant, a duré 8 heures.
+
+Nous souhaitions trouver une alternative, à savoir effectuer ce changement sans que les utilisateurs s'en rendent compte:
+
+- pas d'interruption de service, même sous la forme d'une fenêtre de maintenance planifiée ;
+- pas de dégradation sensible des temps de réponse pendant les opérations.
+
+## Recherche de solution
+
+Ce [thread Stack Overflow](https://stackoverflow.com/questions/54795701/migrating-int-to-bigint-in-postgressql-without-any-downtime/54796046) est un excellent point d'entrée sur le sujet.
+
+Il existe deux solutions :
+
+- soit utiliser une colonne temporaire ;
+- soit utiliser une base de données temporaire, alimentée par une réplication logique.
+
+Nous avons choisi la colonne temporaire, car cette solution peut être mise en œuvre immédiatement dans notre environnement.
+
+En effet, nous utilisons les services d'un [PaaS](https://scalingo.com/), Scalingo, et sommes heureux de ne pas avoir à installer et maintenir nos bases de données. Le service spécifique de la réplication logique n'est pour l'instant pas offert.
+
+Examinons ensemble les étapes de cette [solution](https://tech.coffeemeetsbagel.com/schema-migration-from-int-to-bigint-on-a-massive-table-in-postgres-aabb835c3b84) avec colonne temporaire.
+
+
+### Etape 1 : convertir les données en tâche de fond
+
+La commande `ALTER TABLE foo COLUMN idBigInteger BIG INTEGER` effectue implicitement une conversion des données (cast) de INTEGER à BIGINT.
+
+La [documentation](https://www.postgresql.org/docs/current/sql-altertable.html#notes) mentionne en effet :
+> Changing the type of an existing column will require the entire table and its indexes to be rewritten.
+> As an exception, when changing the type of an existing column and the old type is either binary coercible to the new type a table rewrite is not needed; but any indexes on the affected columns must still be rebuilt.
+> Table and/or index rebuilds may take a significant amount of time for a large table; and will temporarily require as much as double the disk space.
+
+Comme l'opération native pose un verrou exclusif sur la table, il faut attendre sa réécriture complète pour y accéder.
+> An ACCESS EXCLUSIVE lock is held unless explicitly noted.
```

Comment:
C'est intéressant de mentionner la solution "naïve" et pourquoi elle ne répond pas à nos critères, mais c'est étonnant de trouver ça dans une phase où on a déjà "choisi la colonne temporaire" ?
En fait ce paragraphe est redondant avec celui qui commence par "Il est bien sûr possible de modifier le type de cette propriété avec l'instruction native…" plus haut non ?

Reply:
Pas tout à fait. Je cherche à dire qu'il y a plusieurs tâches à effectuer:
- réécriture de la table
- réécriture des index
- recréation des contraintes.

Cela s'applique à la solution native, à celle de l'article original (coffee & bagels), et à la nôtre. 

Je veux :
- expliquer telle solution est plus efficace
- pointer les différences.

Notamment, je veux séparer ce qui était déjà fait et ce que l'on a ajouté  c'est à dire les index et contraintes avant la fenêtre de maintenance.

Ceci dit, ta remarque met en avant le fait que ce n'est pas explicité. Je garde donc ce paragraphe mais

Reply:
"Ce qui était déjà fait" ? Je suis perdu parce qu'on n'utilise rien de la solution `ALTER TABLE … SET TYPE`. Mais bon, si ça te paraît compréhensible, all I can say is 🤷‍♂️